File: apps/marketplace/apps/api/src/adapters/storage.ts
import crypto from 'node:crypto';
import { config } from '../lib/config.js';

export interface StorageAdapter {
  presignPut(key: string, contentType: string, maxBytes: number): Promise<{ url: string; key: string }>;
  publicUrl(key: string): string;
}

class FakeStorage implements StorageAdapter {
  // Dev-mode adapter: returns a deterministic local URL so the mobile app can pretend to upload.
  async presignPut(key: string): Promise<{ url: string; key: string }> {
    return { url: `http://localhost:9000/${config.S3_BUCKET}/${key}`, key };
  }
  publicUrl(key: string): string {
    return `http://localhost:9000/${config.S3_BUCKET}/${key}`;
  }
}

class S3Storage implements StorageAdapter {
  // Minimal S3 v4 signing for PUT presigned URLs without a full SDK dependency.
  async presignPut(key: string, contentType: string): Promise<{ url: string; key: string }> {
    const region = config.S3_REGION;
    const bucket = config.S3_BUCKET;
    const host = config.S3_ENDPOINT
      ? new URL(config.S3_ENDPOINT).host
      : `${bucket}.s3.${region}.amazonaws.com`;
    const date = new Date();
    const amzDate = date.toISOString().replace(/[:-]|\.\d{3}/g, '');
    const dateStamp = amzDate.slice(0, 8);
    const credential = `${config.S3_ACCESS_KEY}/${dateStamp}/${region}/s3/aws4_request`;

    const params = new URLSearchParams({
      'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
      'X-Amz-Credential': credential,
      'X-Amz-Date': amzDate,
      'X-Amz-Expires': '600',
      'X-Amz-SignedHeaders': 'host',
      'Content-Type': contentType,
    });

    const canonicalRequest = [
      'PUT',
      `/${encodeURIComponent(key)}`,
      params.toString(),
      `host:${host}\n`,
      'host',
      'UNSIGNED-PAYLOAD',
    ].join('\n');

    const stringToSign = [
      'AWS4-HMAC-SHA256',
      amzDate,
      `${dateStamp}/${region}/s3/aws4_request`,
      crypto.createHash('sha256').update(canonicalRequest).digest('hex'),
    ].join('\n');

    const kDate = hmac(`AWS4${config.S3_SECRET_KEY}`, dateStamp);
    const kRegion = hmac(kDate, region);
    const kService = hmac(kRegion, 's3');
    const kSigning = hmac(kService, 'aws4_request');
    const signature = hmac(kSigning, stringToSign).toString('hex');

    params.append('X-Amz-Signature', signature);
    const url = `https://${host}/${encodeURIComponent(key)}?${params.toString()}`;
    return { url, key };
  }

  publicUrl(key: string): string {
    if (config.S3_ENDPOINT) return `${config.S3_ENDPOINT}/${config.S3_BUCKET}/${key}`;
    return `https://${config.S3_BUCKET}.s3.${config.S3_REGION}.amazonaws.com/${key}`;
  }
}

function hmac(key: string | Buffer, value: string): Buffer {
  return crypto.createHmac('sha256', key).update(value).digest();
}

export const storage: StorageAdapter = config.S3_FAKE ? new FakeStorage() : new S3Storage();

File: apps/marketplace/apps/api-py/app/adapters/storage.py
"""S3 presigned PUT URL generator (lightweight, no boto3 dependency)."""
from __future__ import annotations

import datetime as dt
import hashlib
import hmac
from urllib.parse import quote, urlencode

from app.config import settings


def _hmac(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode(), hashlib.sha256).digest()


def public_url(key: str) -> str:
    if settings.s3_endpoint:
        return f"{settings.s3_endpoint}/{settings.s3_bucket}/{key}"
    return f"https://{settings.s3_bucket}.s3.{settings.s3_region}.amazonaws.com/{key}"


def presign_put(key: str, content_type: str, expires_seconds: int = 600) -> dict[str, str]:
    """Returns a presigned PUT URL valid for `expires_seconds`."""
    if settings.s3_fake:
        return {"url": public_url(key), "key": key}

    assert settings.s3_access_key and settings.s3_secret_key
    region = settings.s3_region
    host = (
        settings.s3_endpoint.removeprefix("https://").removeprefix("http://")
        if settings.s3_endpoint
        else f"{settings.s3_bucket}.s3.{region}.amazonaws.com"
    )
    now = dt.datetime.now(dt.UTC)
    amz_date = now.strftime("%Y%m%dT%H%M%SZ")
    date_stamp = now.strftime("%Y%m%d")
    credential = f"{settings.s3_access_key}/{date_stamp}/{region}/s3/aws4_request"

    params = {
        "X-Amz-Algorithm": "AWS4-HMAC-SHA256",
        "X-Amz-Credential": credential,
        "X-Amz-Date": amz_date,
        "X-Amz-Expires": str(expires_seconds),
        "X-Amz-SignedHeaders": "host",
        "Content-Type": content_type,
    }
    canonical_query = urlencode(sorted(params.items()), quote_via=quote)
    canonical_request = "\n".join(
        [
            "PUT",
            f"/{quote(key)}",
            canonical_query,
            f"host:{host}\n",
            "host",
            "UNSIGNED-PAYLOAD",
        ]
    )
    string_to_sign = "\n".join(
        [
            "AWS4-HMAC-SHA256",
            amz_date,
            f"{date_stamp}/{region}/s3/aws4_request",
            hashlib.sha256(canonical_request.encode()).hexdigest(),
        ]
    )
    k_date = _hmac(f"AWS4{settings.s3_secret_key}".encode(), date_stamp)
    k_region = _hmac(k_date, region)
    k_service = _hmac(k_region, "s3")
    k_signing = _hmac(k_service, "aws4_request")
    signature = hmac.new(k_signing, string_to_sign.encode(), hashlib.sha256).hexdigest()

    params["X-Amz-Signature"] = signature
    url = f"https://{host}/{quote(key)}?{urlencode(params, quote_via=quote)}"
    return {"url": url, "key": key}
